pageserver/src/tenant.rs

```diff
@@ -1908,16 +1908,20 @@ impl TenantShard {
             .map_err(LoadLocalTimelineError::ResumeDeletion)?;
         }
 
-        // Stash the preloaded tenant manifest, and upload a new manifest if changed.
+        // Upload the tenant manifest.
+        //
+        // This is uploaded unconditionally on every attach. This prevents races where a stale,
+        // still-alive tenant may modify a past manifest, and a future tenant loads it after this
+        // tenant has acted on it. Uploading a new manifest effectively hands over ownership of the
+        // manifest state. See: <https://databricks.atlassian.net/browse/LKB-165>.
         //
         // NB: this must happen after the tenant is fully populated above. In particular the
         // offloaded timelines, which are included in the manifest.
-        {
-            let mut guard = self.remote_tenant_manifest.lock().await;
-            assert!(guard.is_none(), "tenant manifest set before preload"); // first populated here
-            *guard = preload.tenant_manifest;
-        }
-        self.maybe_upload_tenant_manifest().await?;
+        assert!(
+            self.remote_tenant_manifest.lock().await.is_none(),
```

Comment:
this is a bit reliant on the temporary behaviour. It would be nice to still put this into a block, to make the drop more explicit.

Reply:
That seems a bit unnecessary. This behavior is specified in the Rust reference, and if they ever were to make a breaking change for it (seems unlikely) then the `self.maybe_upload_tenant_manifest()` call below wouldn't compile because we're borrowing `self.remote_tenant_manifest`.

Reply:
fair. closing.